File: test_rigid_transform_3D.m
% Generate a random valid rotation matrix
R = orth(rand(3,3)); % random rotation matrix

if det(R) < 0
    [U,S,V] = svd(R);
    V(:,3) = - V(:,3);
    R = V*U';
end

% Generate random translation
t = rand(3,1);

% Generate some random points
n = 10; % number of points
A = rand(3,n);

% Apply transform to get new dataset B
B = R*A + repmat(t, 1, n);

% Recover R and t
[ret_R, ret_t] = rigid_transform_3D(A, B);

% Compare the recovered R and t with the original
B2 = (ret_R*A) + repmat(ret_t, 1, n);

% Find the root mean squared error
err = B2 - B;
err = err .* err;
err = sum(err(:));
rmse = sqrt(err/n);

fprintf("Points A\n")
A

fprintf("Points B\n")
B

fprintf("Ground truth rotation\n")
R

fprintf("Recovered rotation\n")
ret_R

fprintf("Ground truth translation\n")
t

fprintf("Recovered translation\n")
ret_t

fprintf("RMSE: %f\n", rmse);

if rmse < 1e-5
    fprintf("Everything looks good!\n");
else
    fprintf("Hmm something doesn't look right ...\n");
end

File: test_rigid_transform_3D.py
#!/usr/bin/env python3

import numpy as np
import unittest
from scipy.spatial.transform import Rotation
from rigid_transform_3D import rigid_transform_3D


def random_points(n=100):
    return np.random.rand(n, 3)


def random_rotation():
    R = Rotation.from_rotvec(np.random.rand(3)).as_matrix()

    # remove reflection
    # https://en.wikipedia.org/wiki/Kabsch_algorithm
    if np.linalg.det(R) < 0:
        U, _, Vt = np.linalg.svd(R)
        S = np.diag([1, 1, -1])
        R = U @ S @ Vt

    return R


def random_translation():
    return np.random.rand(3, 1)


class Test(unittest.TestCase):
    def test_bad_input(self):
        src_pts = random_points(100)
        dst_pts = random_points(50)

        with self.assertRaises(AssertionError):
            rigid_transform_3D(src_pts, dst_pts)

    def test_Nx3(self):
        pts = random_points(100)
        R, t = rigid_transform_3D(pts, pts)

        self.assertAlmostEqual(np.sum((R - np.eye(3)) ** 2), 0.0)
        self.assertAlmostEqual(np.sum(t**2), 0.0)

    def test_3xN(self):
        pts = random_points(100).T
        R, t = rigid_transform_3D(pts, pts)

        self.assertAlmostEqual(np.sum((R - np.eye(3)) ** 2), 0.0)
        self.assertAlmostEqual(np.sum(t**2), 0.0)

    def test(self):
        R = random_rotation()
        t = random_translation()
        src = random_points()
        dst = R @ src.T + t

        # Recover R and t
        ret_R, ret_t = rigid_transform_3D(src, dst.T)
        dst2 = ret_R @ src.T + ret_t

        rmse = np.sqrt(np.mean(((dst - dst2) ** 2).flatten()))

        print("Ground truth rotation")
        print(R)

        print("Recovered rotation")
        print(ret_R)
        print("")

        print("Ground truth translation: ", t.flatten())
        print("Recovered translation: ", ret_t.flatten())
        print("")
        print("RMSE:", rmse)

        if rmse < 1e-5:
            print("Everything looks good!")
        else:
            print("Hmm something doesn't look right ...")

        self.assertAlmostEqual(rmse, 0.0)


if __name__ == "__main__":
    unittest.main()
